File: binding/src/network/libqsslcertificate.cpp
#include <QByteArray>
#include <QDateTime>
#include <QDebug>
#include <QIODevice>
#include <QSslCertificate>
#include <QSslKey>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qsslcertificate.h>
#include "libqsslcertificate.h"
#include "libqsslcertificate.hxx"

QSslCertificate* QSslCertificate_new(QIODevice* device) {
	 return new QSslCertificate(device);
}

QSslCertificate* QSslCertificate_new2() {
	 return new QSslCertificate();
}

QSslCertificate* QSslCertificate_new3(const QSslCertificate* other) {
	 return new QSslCertificate(*other);
}

QSslCertificate* QSslCertificate_new4(QIODevice* device, int format) {
	 return new QSslCertificate(device, static_cast<QSsl::EncodingFormat>(format));
}

QSslCertificate* QSslCertificate_new5(const libqt_string data) {
	 return new QSslCertificate(QByteArray(data.data, data.len));
}

QSslCertificate* QSslCertificate_new6(const libqt_string data, int format) {
	 return new QSslCertificate(QByteArray(data.data, data.len), static_cast<QSsl::EncodingFormat>(format));
}

void QSslCertificate_OperatorAssign(QSslCertificate* self, const QSslCertificate* other) {
	self->operator=(*other);
}

void QSslCertificate_Swap(QSslCertificate* self, QSslCertificate* other) {
	self->swap(*other);
}

bool QSslCertificate_OperatorEqual(const QSslCertificate* self, const QSslCertificate* other) {
	return self->operator==(*other);
}

bool QSslCertificate_OperatorNotEqual(const QSslCertificate* self, const QSslCertificate* other) {
	return self->operator!=(*other);
}

bool QSslCertificate_IsNull(const QSslCertificate* self) {
	return self->isNull();
}

bool QSslCertificate_IsBlacklisted(const QSslCertificate* self) {
	return self->isBlacklisted();
}

bool QSslCertificate_IsSelfSigned(const QSslCertificate* self) {
	return self->isSelfSigned();
}

void QSslCertificate_Clear(QSslCertificate* self) {
	self->clear();
}

libqt_string QSslCertificate_Version(const QSslCertificate* self) {
	QByteArray _qb = self->version();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSslCertificate_SerialNumber(const QSslCertificate* self) {
	QByteArray _qb = self->serialNumber();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSslCertificate_Digest(const QSslCertificate* self) {
	QByteArray _qb = self->digest();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QSslCertificate_IssuerInfo(const QSslCertificate* self, int info) {
	auto _ret = self->issuerInfo(static_cast<QSslCertificate::SubjectInfo>(info));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_list QSslCertificate_IssuerInfo2(const QSslCertificate* self, const libqt_string attribute) {
	auto _ret = self->issuerInfo(QByteArray(attribute.data, attribute.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_list QSslCertificate_SubjectInfo(const QSslCertificate* self, int info) {
	auto _ret = self->subjectInfo(static_cast<QSslCertificate::SubjectInfo>(info));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_list QSslCertificate_SubjectInfo2(const QSslCertificate* self, const libqt_string attribute) {
	auto _ret = self->subjectInfo(QByteArray(attribute.data, attribute.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_string QSslCertificate_IssuerDisplayName(const QSslCertificate* self) {
	QString _ret = self->issuerDisplayName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSslCertificate_SubjectDisplayName(const QSslCertificate* self) {
	QString _ret = self->subjectDisplayName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QSslCertificate_SubjectInfoAttributes(const QSslCertificate* self) {
	auto _ret = self->subjectInfoAttributes();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QSslCertificate_IssuerInfoAttributes(const QSslCertificate* self) {
	auto _ret = self->issuerInfoAttributes();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_map QSslCertificate_SubjectAlternativeNames(const QSslCertificate* self) {
	auto _ret = self->subjectAlternativeNames();
	libqt_map _map;
	_map.len = _ret.size();
	_map.keys = nullptr;
	_map.values = nullptr;
	return _map;
}

QDateTime* QSslCertificate_EffectiveDate(const QSslCertificate* self) {
	return new QDateTime(self->effectiveDate());
}

QDateTime* QSslCertificate_ExpiryDate(const QSslCertificate* self) {
	return new QDateTime(self->expiryDate());
}

QSslKey* QSslCertificate_PublicKey(const QSslCertificate* self) {
	return new QSslKey(self->publicKey());
}

libqt_list QSslCertificate_Extensions(const QSslCertificate* self) {
	auto _ret = self->extensions();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_string QSslCertificate_ToPem(const QSslCertificate* self) {
	QByteArray _qb = self->toPem();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSslCertificate_ToDer(const QSslCertificate* self) {
	QByteArray _qb = self->toDer();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSslCertificate_ToText(const QSslCertificate* self) {
	QString _ret = self->toText();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QSslCertificate_FromPath(const libqt_string path) {
	auto _ret = QSslCertificate::fromPath(QString::fromUtf8(path.data, path.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QSslCertificate_FromDevice(QIODevice* device) {
	auto _ret = QSslCertificate::fromDevice(device);
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QSslCertificate_FromData(const libqt_string data) {
	auto _ret = QSslCertificate::fromData(QByteArray(data.data, data.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QSslCertificate_Verify(const libqt_list certificateChain) {
	auto _ret = QSslCertificate::verify(QList<QSslCertificate>());
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

bool QSslCertificate_ImportPkcs12(QIODevice* device, QSslKey* key, QSslCertificate* cert) {
	return QSslCertificate::importPkcs12(device, key, cert);
}

void* QSslCertificate_Handle(const QSslCertificate* self) {
	return self->handle();
}

libqt_string QSslCertificate_Digest1(const QSslCertificate* self, int algorithm) {
	QByteArray _qb = self->digest(static_cast<QCryptographicHash::Algorithm>(algorithm));
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QSslCertificate_FromPath2(const libqt_string path, int format) {
	auto _ret = QSslCertificate::fromPath(QString::fromUtf8(path.data, path.len), static_cast<QSsl::EncodingFormat>(format));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QSslCertificate_FromPath3(const libqt_string path, int format, int syntax) {
	auto _ret = QSslCertificate::fromPath(QString::fromUtf8(path.data, path.len), static_cast<QSsl::EncodingFormat>(format), static_cast<QSslCertificate::PatternSyntax>(syntax));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QSslCertificate_FromDevice2(QIODevice* device, int format) {
	auto _ret = QSslCertificate::fromDevice(device, static_cast<QSsl::EncodingFormat>(format));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QSslCertificate_FromData2(const libqt_string data, int format) {
	auto _ret = QSslCertificate::fromData(QByteArray(data.data, data.len), static_cast<QSsl::EncodingFormat>(format));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QSslCertificate_Verify2(const libqt_list certificateChain, const libqt_string hostName) {
	auto _ret = QSslCertificate::verify(QList<QSslCertificate>(), QString::fromUtf8(hostName.data, hostName.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

bool QSslCertificate_ImportPkcs124(QIODevice* device, QSslKey* key, QSslCertificate* cert, libqt_list caCertificates) {
	return QSslCertificate::importPkcs12(device, key, cert, QList<QSslCertificate>());
}

bool QSslCertificate_ImportPkcs125(QIODevice* device, QSslKey* key, QSslCertificate* cert, libqt_list caCertificates, const libqt_string passPhrase) {
	return QSslCertificate::importPkcs12(device, key, cert, QList<QSslCertificate>(), QByteArray(passPhrase.data, passPhrase.len));
}

void QSslCertificate_Delete(QSslCertificate* self) {
    delete self;
}

File: binding/include/network/libqsslcertificate.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsslcertificate.html
pub const qsslcertificate = struct {

    /// New constructs a new QSslCertificate object.
    pub fn New(device: ?*anyopaque) QtC.QSslCertificate {
        return qtc.QSslCertificate_new(@ptrCast(device));
    }


    /// New2 constructs a new QSslCertificate object.
    pub fn New2() QtC.QSslCertificate {
        return qtc.QSslCertificate_new2();
    }


    /// New3 constructs a new QSslCertificate object.
    pub fn New3(other: ?*anyopaque) QtC.QSslCertificate {
        return qtc.QSslCertificate_new3(@ptrCast(other));
    }


    /// New4 constructs a new QSslCertificate object.
    pub fn New4(device: ?*anyopaque, format: i32) QtC.QSslCertificate {
        return qtc.QSslCertificate_new4(@ptrCast(device), @intCast(format));
    }


    /// New5 constructs a new QSslCertificate object.
    pub fn New5(data: []u8) QtC.QSslCertificate {
        const data_str = qtc.libqt_string{
    .len = data.len,
    .data = data.ptr,
};
return qtc.QSslCertificate_new5(data_str);
    }


    /// New6 constructs a new QSslCertificate object.
    pub fn New6(data: []u8, format: i32) QtC.QSslCertificate {
        const data_str = qtc.libqt_string{
    .len = data.len,
    .data = data.ptr,
};
return qtc.QSslCertificate_new6(data_str, @intCast(format));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSslCertificate_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSslCertificate_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QSslCertificate_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QSslCertificate_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsNull(self: ?*anyopaque, ) bool {
        return qtc.QSslCertificate_IsNull(@ptrCast(self));
    }

    pub fn IsBlacklisted(self: ?*anyopaque, ) bool {
        return qtc.QSslCertificate_IsBlacklisted(@ptrCast(self));
    }

    pub fn IsSelfSigned(self: ?*anyopaque, ) bool {
        return qtc.QSslCertificate_IsSelfSigned(@ptrCast(self));
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QSslCertificate_Clear(@ptrCast(self));
    }

    pub fn Version(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QSslCertificate_Version(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qsslcertificate.Version: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn SerialNumber(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QSslCertificate_SerialNumber(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qsslcertificate.SerialNumber: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn Digest(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QSslCertificate_Digest(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qsslcertificate.Digest: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn IssuerInfo(self: ?*anyopaque, info: i32) []const u8 {
        return qtc.QSslCertificate_IssuerInfo(@ptrCast(self), @intCast(info));
    }

    pub fn IssuerInfo2(self: ?*anyopaque, attribute: []u8) []const u8 {
        const attribute_str = qtc.libqt_string{
    .len = attribute.len,
    .data = attribute.ptr,
};
return qtc.QSslCertificate_IssuerInfo2(@ptrCast(self), attribute_str);
    }

    pub fn SubjectInfo(self: ?*anyopaque, info: i32) []const u8 {
        return qtc.QSslCertificate_SubjectInfo(@ptrCast(self), @intCast(info));
    }

    pub fn SubjectInfo2(self: ?*anyopaque, attribute: []u8) []const u8 {
        const attribute_str = qtc.libqt_string{
    .len = attribute.len,
    .data = attribute.ptr,
};
return qtc.QSslCertificate_SubjectInfo2(@ptrCast(self), attribute_str);
    }

    pub fn IssuerDisplayName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSslCertificate_IssuerDisplayName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsslcertificate.IssuerDisplayName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SubjectDisplayName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSslCertificate_SubjectDisplayName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsslcertificate.SubjectDisplayName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SubjectInfoAttributes(self: ?*anyopaque, ) []const u8 {
        return qtc.QSslCertificate_SubjectInfoAttributes(@ptrCast(self));
    }

    pub fn IssuerInfoAttributes(self: ?*anyopaque, ) []const u8 {
        return qtc.QSslCertificate_IssuerInfoAttributes(@ptrCast(self));
    }

    pub fn SubjectAlternativeNames(self: ?*anyopaque, ) QMultiMap<QSsl::AlternativeNameEntryType, QString> {
        return qtc.QSslCertificate_SubjectAlternativeNames(@ptrCast(self));
    }

    pub fn EffectiveDate(self: ?*anyopaque, ) QtC.QDateTime {
        return qtc.QSslCertificate_EffectiveDate(@ptrCast(self));
    }

    pub fn ExpiryDate(self: ?*anyopaque, ) QtC.QDateTime {
        return qtc.QSslCertificate_ExpiryDate(@ptrCast(self));
    }

    pub fn PublicKey(self: ?*anyopaque, ) QtC.QSslKey {
        return qtc.QSslCertificate_PublicKey(@ptrCast(self));
    }

    pub fn Extensions(self: ?*anyopaque, ) []const u8 {
        return qtc.QSslCertificate_Extensions(@ptrCast(self));
    }

    pub fn ToPem(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QSslCertificate_ToPem(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qsslcertificate.ToPem: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn ToDer(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QSslCertificate_ToDer(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qsslcertificate.ToDer: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn ToText(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSslCertificate_ToText(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsslcertificate.ToText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn FromPath(path: []const u8) []const u8 {
        const path_str = qtc.libqt_string{
    .len = path.len,
    .data = path.ptr,
};
return qtc.QSslCertificate_FromPath(path_str);
    }

    pub fn FromDevice(device: ?*anyopaque) []const u8 {
        return qtc.QSslCertificate_FromDevice(@ptrCast(device));
    }

    pub fn FromData(data: []u8) []const u8 {
        const data_str = qtc.libqt_string{
    .len = data.len,
    .data = data.ptr,
};
return qtc.QSslCertificate_FromData(data_str);
    }

    pub fn Verify(certificateChain: []const u8) []const u8 {
        return qtc.QSslCertificate_Verify(@ptrCast(certificateChain));
    }

    pub fn ImportPkcs12(device: ?*anyopaque, key: ?*anyopaque, cert: ?*anyopaque) bool {
        return qtc.QSslCertificate_ImportPkcs12(@ptrCast(device), @ptrCast(key), @ptrCast(cert));
    }

    pub fn Handle(self: ?*anyopaque, ) ?*anyopaque {
        return qtc.QSslCertificate_Handle(@ptrCast(self));
    }

    pub fn Digest1(self: ?*anyopaque, algorithm: i32, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QSslCertificate_Digest1(@ptrCast(self), @intCast(algorithm));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qsslcertificate.Digest1: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn FromPath2(path: []const u8, format: i32) []const u8 {
        const path_str = qtc.libqt_string{
    .len = path.len,
    .data = path.ptr,
};
return qtc.QSslCertificate_FromPath2(path_str, @intCast(format));
    }

    pub fn FromPath3(path: []const u8, format: i32, syntax: i32) []const u8 {
        const path_str = qtc.libqt_string{
    .len = path.len,
    .data = path.ptr,
};
return qtc.QSslCertificate_FromPath3(path_str, @intCast(format), @intCast(syntax));
    }

    pub fn FromDevice2(device: ?*anyopaque, format: i32) []const u8 {
        return qtc.QSslCertificate_FromDevice2(@ptrCast(device), @intCast(format));
    }

    pub fn FromData2(data: []u8, format: i32) []const u8 {
        const data_str = qtc.libqt_string{
    .len = data.len,
    .data = data.ptr,
};
return qtc.QSslCertificate_FromData2(data_str, @intCast(format));
    }

    pub fn Verify2(certificateChain: []const u8, hostName: []const u8) []const u8 {
        const hostName_str = qtc.libqt_string{
    .len = hostName.len,
    .data = hostName.ptr,
};
return qtc.QSslCertificate_Verify2(@ptrCast(certificateChain), hostName_str);
    }

    pub fn ImportPkcs124(device: ?*anyopaque, key: ?*anyopaque, cert: ?*anyopaque, caCertificates: []const u8) bool {
        return qtc.QSslCertificate_ImportPkcs124(@ptrCast(device), @ptrCast(key), @ptrCast(cert), @ptrCast(caCertificates));
    }

    pub fn ImportPkcs125(device: ?*anyopaque, key: ?*anyopaque, cert: ?*anyopaque, caCertificates: []const u8, passPhrase: []u8) bool {
        const passPhrase_str = qtc.libqt_string{
    .len = passPhrase.len,
    .data = passPhrase.ptr,
};
return qtc.QSslCertificate_ImportPkcs125(@ptrCast(device), @ptrCast(key), @ptrCast(cert), @ptrCast(caCertificates), passPhrase_str);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSslCertificate_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/sslcertificate.html#types
pub const enums = struct {
    pub const SubjectInfo = enum {
        pub const Organization: i32 = 0;
        pub const CommonName: i32 = 1;
        pub const LocalityName: i32 = 2;
        pub const OrganizationalUnitName: i32 = 3;
        pub const CountryName: i32 = 4;
        pub const StateOrProvinceName: i32 = 5;
        pub const DistinguishedNameQualifier: i32 = 6;
        pub const SerialNumber: i32 = 7;
        pub const EmailAddress: i32 = 8;
    };

    pub const PatternSyntax = enum {
        pub const RegularExpression: i32 = 0;
        pub const Wildcard: i32 = 1;
        pub const FixedString: i32 = 2;
    };

};
